File: src/cli/verify.ts
// `cappu verify`: check the jars currently in the lib directories against the
// SHA-256 sums in cappu-lock.json. Read-only; exits non-zero on any mismatch
// or missing jar.

import type { CappuConfig } from "../config.ts";
import { verifyInstalled } from "../install.ts";
import { emitAnnotation } from "./annotations.ts";

export function runVerify(config: CappuConfig): never {
  const result = verifyInstalled(config);
  if (!result.fromLock) {
    // A dep-free project with no lock is vacuously verified (matching
    // `install --locked`), not an error.
    const declared = Object.values(config.dependencies).some(m => Object.keys(m).length > 0);
    if (!declared) {
      process.stderr.write("0 ok, 0 modified, 0 missing\n");
      process.exit(0);
    }
    process.stderr.write(
      "cappu: no cappu-lock.json to verify against; run `cappu install` first\n",
    );
    emitAnnotation("error", "no cappu-lock.json to verify against; run `cappu install` first");
    process.exit(1);
  }
  for (const id of result.modified) {
    process.stderr.write(`error: ${id}: installed jar does not match cappu-lock.json\n`);
    emitAnnotation("error", `${id}: installed jar does not match cappu-lock.json`);
  }
  for (const id of result.missing) {
    process.stderr.write(`error: ${id}: locked but not installed (run \`cappu install\`)\n`);
    emitAnnotation("error", `${id}: locked but not installed (run \`cappu install\`)`);
  }
  process.stderr.write(
    `${result.ok.length} ok, ${result.modified.length} modified, ${result.missing.length} missing\n`,
  );
  process.exit(result.modified.length + result.missing.length > 0 ? 1 : 0);
}

File: togo/internal/cli/verify.go
package cli

import (
	"fmt"
	"os"

	"github.com/nikeee/cappu/internal/config"
	"github.com/nikeee/cappu/internal/lockfile"
)

// RunVerify handles `cappu verify`: check the jars currently in the lib
// directories against the SHA-256 sums in cappu-lock.json. Read-only; exits
// non-zero on any mismatch or missing jar. Port of src/cli/verify.ts.
func RunVerify(cfg *config.Config) int {
	result := lockfile.VerifyInstalled(cfg)
	if !result.FromLock {
		fmt.Fprintln(os.Stderr, "cappu: no cappu-lock.json to verify against; run `cappu install` first")
		return 1
	}
	for _, id := range result.Modified {
		fmt.Fprintf(os.Stderr, "error: %s: installed jar does not match cappu-lock.json\n", id)
	}
	for _, id := range result.Missing {
		fmt.Fprintf(os.Stderr, "error: %s: locked but not installed (run `cappu install`)\n", id)
	}
	fmt.Fprintf(os.Stderr, "%d ok, %d modified, %d missing\n",
		len(result.OK), len(result.Modified), len(result.Missing))
	if len(result.Modified)+len(result.Missing) > 0 {
		return 1
	}
	return 0
}
